assignments/serde-lifetimes.adoc

```diff
@@ -0,0 +1,148 @@
+[source,rust]
+----
+// here `"#` would not terminate the string
+r##"
+    {
+        "color": "#ff00ff"
+    }
+"##
+----
+====
+
+== Tasks
+
+=== 1. Serialization and Deserialization using `String` s
+
+We used `todo!()` macros to mark places where you should put code to make the program run. Look at link:https://docs.rs/serde_json/latest/serde_json/#functions[`serde_json`] api for help.
+
+.Hint
+[%collapsible]
+====
+Serde comes with two traits: `Serializable` and `Deserializable`. These traits can be `derive` d for your `struct` or `enum` types. Other `serde-*` crates use these traits to convert our data type from and to corresponding representation (`serde-json` to JSON, `serde-yaml` to YAML, etc.).
+====
+
+=== 2. Deserialization using `str` s
+
+Notice that when we deserialize data we created a separate `String` to represent a `name` for our blog post. If our data type has a lot of fields in it we may end up allocating many-many Strings every time we receive data from our API.
+
+To avoid it we often can use string slices to refer to fragments of the original JSON string.
+
+Try changing or blog post type to the following:
+
+[source,rust]
+----
+struct BlogPost {
+    id: u32,
+
+    title: &str,
+}
+----
+
+The compiler will ask you to provide a lifetime type annotation.
+
+*Q: Can your `title` be `&'static str`? Why or why not?*
+
+Use lifetime annotations and change the code inside `main` if necessary.
+
+*Q: If you made a change, why?*
+
+=== 3. Use copy-on-write.
+
+Let's put the word _Rust_ into quotes in our blog post title. Double quotes have to be escaped in JSON, so try changing the title to `"Hello, \"Rust\""`.
+
+Run your program, and you'll get an error:
+
+[source,rust]
+----
+Error("invalid type: string \"Hello, \\\"Rust\\\"\", expected a borrowed string", ... )
+----
+
+link:https://doc.rust-lang.org/std/borrow/enum.Cow.html[`std::borrow::Cow`] can be used to represent data that can be borrowed (referenced) or owned. It provides Copy-on-Write semantics.
+
+Change `title` type to `Cow<'a, str>`. The error should disappear.
+
+=== 4. Teach Serde to not do unnecessary allocations.
+
+Let's observe `Cow` behavior. Add this to your `main` function after deserialization step:
+
+[source,rust]
+----
+println!("is borrowed = {}", matches!(post.title, Cow::Borrowed(_)));
+----
+
+Now change the title of the blog post back to `"Hello, Rust"` - without double quotes, and run the program again.
+
+Turns out Serde creates an owned copy of the data anyway. To avoid that use link:https://serde.rs/field-attrs.html#borrow[`#[serde(borrow)\]`] macro.
+
+Get to the point when your program prints: `is borrowed = true`
+
+== Questions after this exercise
+
+ 1. Why is there a need for separate `serde` and `serde-json` crates?
+ 2. Why do we need to put a lifetime annotation for a field of '&str' type?
+ 3. What about `Cow` type?
```

Comment:
Maybe reframe this question more clearly? I am not sure what you are asking about.